File: src/1020_number_of_enclaves.js
function numEnclaves(grid) {
    const rows = grid.length;
    const cols = grid[0].length;
    const dfs = (row, col) => {
        if (row >= 0 && row < rows && col >= 0 && col < cols && grid[row][col] === 1) {
            grid[row][col] = 0;
            dfs(row + 1, col);
            dfs(row - 1, col);
            dfs(row, col + 1);
            dfs(row, col - 1);
        }
    };
    for (let row = 0; row < rows; row++) {
        dfs(row, 0);
        dfs(row, cols - 1);
    }
    for (let col = 0; col < cols; col++) {
        dfs(0, col);
        dfs(rows - 1, col);
    }
    let total = 0;
    for (let row = 0; row < rows; row++) {
        for (let col = 0; col < cols; col++) {
            if (grid[row][col] === 1) total++;
        }
    }
    return total;
}

module.exports = { numEnclaves };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(numEnclaves([[0, 0, 0, 0], [1, 0, 1, 0], [0, 1, 1, 0], [0, 0, 0, 0]]), 3);
    assert.deepStrictEqual(numEnclaves([[0, 1, 1, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0, 0, 0, 0]]), 0);
}

File: src/1020_number_of_enclaves.rs
struct Solution;

impl Solution {
    fn enclave_dfs(grid: &mut Vec<Vec<i32>>, row: i32, col: i32, rows: i32, cols: i32) {
        if (0..rows).contains(&row)
            && (0..cols).contains(&col)
            && grid[row as usize][col as usize] == 1
        {
            grid[row as usize][col as usize] = 0;
            Self::enclave_dfs(grid, row + 1, col, rows, cols);
            Self::enclave_dfs(grid, row - 1, col, rows, cols);
            Self::enclave_dfs(grid, row, col + 1, rows, cols);
            Self::enclave_dfs(grid, row, col - 1, rows, cols);
        }
    }

    pub fn num_enclaves(mut grid: Vec<Vec<i32>>) -> i32 {
        let rows = grid.len() as i32;
        let cols = grid[0].len() as i32;

        for row in 0..rows {
            Self::enclave_dfs(&mut grid, row, 0, rows, cols);
            Self::enclave_dfs(&mut grid, row, cols - 1, rows, cols);
        }

        for col in 0..cols {
            Self::enclave_dfs(&mut grid, 0, col, rows, cols);
            Self::enclave_dfs(&mut grid, rows - 1, col, rows, cols);
        }

        let mut total = 0;
        for row in 0..rows {
            for col in 0..cols {
                if grid[row as usize][col as usize] == 1 {
                    total += 1;
                }
            }
        }
        total
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::num_enclaves(vec![
                vec![0, 0, 0, 0],
                vec![1, 0, 1, 0],
                vec![0, 1, 1, 0],
                vec![0, 0, 0, 0]
            ]),
            3
        );
        assert_eq!(
            Solution::num_enclaves(vec![
                vec![0, 1, 1, 0],
                vec![0, 0, 1, 0],
                vec![0, 0, 1, 0],
                vec![0, 0, 0, 0]
            ]),
            0
        );
    }
}
